### ISSUE TITLE
Batch Dataloader and Dataset

### ISSUE THREAD

**Issue by EvenOldridge** [pytorch/pytorch#21645]
## 🚀 Feature
A dataloader and dataset that operate at the batch level, rather than the item level, pulling batches from contiguous blocks of memory and avoiding random access patterns in the dataloader.

## Motivation
Loading data item by item and coallating into a batch is very inefficient, particularly in the case of tabular or text data where the items are small.  This is compounded further when you want to use large batch sizes.  By pre shuffling the data each epoch (when required) we can grab each batch as a single read from contiguous memory.  This much faster and scales better with batch size, removing the necessity of multiprocessing, which adds complexity in the form of bus errors when not enough shared memory is available (https://github.com/pytorch/pytorch/issues/5040), CUDA init issues when forking (https://github.com/pytorch/pytorch/issues/4377), etc.  This forking issue was one of my original motivations as it solves the issue of using the dataloader in conjunction with RAPIDS or any other code that calls CUDA before the dataloader workers are forked.  It should also solve the issue on windows with the speed of dataloaders, at least for tabular and text data,  (https://github.com/pytorch/pytorch/issues/12831) as spawning is not necessary.  

Using the proposed method results in better GPU utilization, and better throughput when training in the tests on tabular data that I've run.  With no multiprocessing I've measured a 5-15% improvement* in throughput over an 8 worker vanilla dataloader (more were tried but it maxed out at 8).  I've also been able to increase batch sizes for tabular data into the 800K+ range with no loss of accuracy and get a 2x performance improvement over the best multiprocessor dataloader I could run without running into bus error issues that cropped up with large batch sizes.

*depends on tensor and batch size

## Pitch

I've created source for a batch dataloader and batch dataset modelled after their vanilla counterparts and would love to see it integrated into the PyTorch repo.  Usage is similar, and I've tried to stick to the pytorch variable naming and formatting.

Code can be found here: https://github.com/rapidsai/dataloaders/tree/master/pytorch/batch_dataloader

It should hopefully be ready to go; I've tested it with both base pytorch and with ignite, but more eyes on it would definitely be beneficial, particularly in use cases beyond tabular like text or small images.  It should be applicable to anyone who isn't doing large images or a lot of image augmentation.  It's undergone an internal (NVidia) review of @ptrblck who was immensely helpful in refining it and @ngimel who reviewed the codebase and had helpful suggestions regarding memory pinning.  

I'm happy to work with the team to create test cases similar to those for dataset and dataloader and would love feedback on it.

## Alternatives

One possible solution to the CUDA Init before fork issue is to spawn, however as seen in windows this is significantly slower and I had trouble getting it working.   

## Additional context

I'm also working on versions of this that work with larger than CPU memory datasets and on a version that works in GPU memory doing a 0-copy transform of a rapids cudf dataframe via dlpack.

**Comment by fmassa** [comment]
I believe this is being implemented in https://github.com/pytorch/pytorch/pull/19228

Let us know if there is anything else your proposal that is not present in #19228

**Comment by ngimel** [comment]
Correct me if I'm wrong, but it looks to me that #19228 still reads data item by item, even in batched mode https://github.com/pytorch/pytorch/blob/1834c27184387175b39722b750c153a37754766c/torch/utils/data/_utils/fetch.py, which is what @EvenOldridge wants to avoid. (this is not saying that #19228 could not be changed to accomodate fetching a full batch at once). Also, looks like #19228 is stalled?

**Comment by fmassa** [comment]
I don't think #19228 is stale, and @SsnL is still working on getting it merged.
Also, I was under the impression that batched mode was added explicitly to support fetching a full batch at once, on systems where performing a query is expensive. But maybe @SsnL can give more details on it?

**Comment by fmassa** [comment]
Also, @ngimel the batched fetching is done in https://github.com/pytorch/pytorch/blob/1834c27184387175b39722b750c153a37754766c/torch/utils/data/_utils/fetch.py#L46 I believe

**Comment by SsnL** [comment]
Yep I can clarify a bit. This indeed can be solved by #19228 (which is not stale). To do this, disable automatic batching and handle collation oneself in an IterableDataset’s __iter__.

**Comment by ngimel** [comment]
Thanks for the clarification!

**Comment by SsnL** [comment]
The auto-collation in fetch.py is meant to be used by cases where batch size is not data dependent and chunk reading is not too beneficial, I.e. majority of small to medium size training. It is provided for users to easily implement custom dataset without figuring out complicated efficient batching code themselves. After #19228 , setting batch_size=None (and batch_sampler=None) disables autobatching (auto collation) and advanced use cases can handle batching themselves.

**Comment by EvenOldridge** [comment]
@SsnL Some very cool work.  I looked into the multiprocessing side of dataloaders but when I found it wasn't necessary for my use case of tabular data I went with a simpler approach.  The iterable style looks particularly useful for the larger than memory dataloader that I'm currently working on.

Can you clarify the below further:

>To do this, disable automatic batching and handle collation oneself in an IterableDataset’s iter. 

I'm trying to understand how a user would take advantage of batching with a single read from contiguous memory rather than the item by item followed by coallation.  Is this something they would have to write for themselves?

If so then I would argue that while you can technically achieve the same outcome, the majority of users won't every see those benefits.  I think the performance improvements and removal of multiprocessing requirements make an example of this desirable.  Ideally we can even merge this in so that it's the default when batch size is specified.

**Comment by EvenOldridge** [comment]
@SsnL I understand you're probably very busy with your PR but I'd like to make sure this doesn't just disappear.  Can you confirm whether or not your dataloader reads batches as contiguous blocks of memory.  From what's being described it's possible under the iterable dataloader but not built in?  

If not maybe we can work together to add that functionality or I can follow up on this once your PR is added and add that functionality myself.  The improved GPU utilization, faster throughput, and lack of multiprocessor dependency make it a better option than per item dataloading in most cases.

**Comment by SsnL** [comment]
@EvenOldridge Hi sorry for the late reply. I'm not at FB so at best I am working on PyTorch parttime. 

#19228 solves this in the sense that auto-collation becomes optional. With either the old map-style dataset, or the new iterable-style dataset, one can do batching themselves within the dataset code, as what you have done in your repo. Then simply disabling multiprocess loading achieves what your asked for. (Also, there is never a dependency on multiprocessing.)

You are right in that after #19228 it does still require user code, but it requires much less. Something like just these lines should suffices https://github.com/rapidsai/dataloaders/blob/master/pytorch/batch_dataloader/batch_dataset.py#L27-L60. This is similar or less amount of code to most pytorch dataset use cases.

I'm inclining to not include this particular dataset class in core. The general philosophy of pytorch data loading is that we provide generic and easy-to-use-and-extend interface for users, who only needs to provide a custom dataset object. Admittedly, the `TensorBatchDataset` is useful for the use case described. But my feeling is that it might be too narrow to be part of the core library that strikes to be general. E.g., what if the data sample has some complicated data structure than just a list of tensors? This is my $0.02.

**Comment by EvenOldridge** [comment]
I think you might be missing a few key points.  

>there is never a dependency on multiprocessing
The dependency is there if you want to achieve performance.  Benchmarking non-multi vs multiprocessors in my problem domain shows a 3-5x difference in throughput.  So while you can disable it you're performance tanks.

> my feeling is that it might be too narrow to be part of the core library that strikes to be general. E.g., what if the data sample has some complicated data structure than just a list of tensors? 

The submission is for a batch dataset and batch dataloader class that is extended to tensors in TensorBatchDataset.  This is the exact same method that dataset and dataloader use.  If you want a dataset more complicated that just a list of tensors it should still be possible to store them in contiguous memory and load them in one read.

Based on the context above and the added flexibility your pr provides I can however understand not including this in core, particularly in it's current format.  

When your PR is merged would it be helpful for me to provide an example like TensorBatchDataset to show users how to do this?  My concern is that they're going to stick with the default option when this method allows for higher throughput than multiprocessing without doing any multiprocessing.  

It seems a shame to just expect the user to figure this out on their own, especially considering the complexity of the dataloader.

**Comment by wconnell** [comment]
This was very helpful to come across @EvenOldridge. 

> It seems a shame to just expect the user to figure this out on their own, especially considering the complexity of the dataloader.

Agreed. An example would be quite helpful in the least. With increasing tabular data use cases I don't think it is wild to consider including this type of class in core. The default automatic batching and collation setup has its own data type biases imo.

**Comment by thebabellibrarybot** [comment]
can anyone write stuff here?
